app/controllers/crypto_address_for_channels_controller.rb

```diff
@@ -0,0 +1,114 @@
+require "eth"
+require "rbnacl"
+require "base58"
+
+class CryptoAddressForChannelsController < ApplicationController
+  include Eth
+  before_action :authenticate_publisher!
+
+  def index
+    current_channel = current_publisher.channels.find(params[:channel_id])
+    @crypto_addresses_for_channel = CryptoAddressForChannel.where(channel: current_channel)
+    render(json: @crypto_addresses_for_channel)
+  end
+
+  def generate_nonce
+    set = [('a'..'z'), ('A'..'Z'), (0..9)].map(&:to_a).flatten
+    nonce = (0...32).map { set[rand(set.length)] }.join
+    Rails.cache.write(nonce, true)
+    render json: {nonce: nonce}
+  end
+
+  def create
+    signature = params[:transaction_signature]
+    account_address = params[:account_address]
+    chain = params[:chain]
+    message = params[:message]
+    current_channel = current_publisher.channels.find(params[:channel_id])
+    @errors = []
+
+    # check that the message is a valid nonce and delete after use
+    valid_message = if Rails.cache.read(message)
+      !!Rails.cache.delete(message)
```

Comment:
Probably a race condition allowing multiple use here, but no real impact so I'd ignore it.

Reply:
If this cache is reused across all channels then this could be a problem here where a replay attack could be used to perform an authentication bypass that allows one creator to steal another creators signed message and replay it before they are able to submit it which would then link the address to the wrong creator.

This can be avoided by ideally linking the nonce to a specific creator such that even if the message were relayed we'd only link the address to the original account that requested the nonce.

Reply:
I've updated the secureRandom to use uuid instead of hex.  Rather than adding the channel id to the nonce, which is visible from the URL, I'm storing the publisher ID with the nonce on the back end, so no one but the correct logged in publisher can use the nonce.